source/controllers/include/controllers/impedance/Impedance.hpp

```diff
@@ -0,0 +1,133 @@
+#include "state_representation/Parameters/Parameter.hpp"
+#include <eigen3/Eigen/Core>
+#include <eigen3/Eigen/Dense>
+
+namespace controllers {
+namespace impedance {
+/**
+ * @class Impedance
+ * @brief Definition of an impedance controller in either joint or task space
+ * @tparam S the space of the controller either CartesianState or JointState
+ */
+template <class S>
+class Impedance : Controller<S, S> {
+private:
+  std::shared_ptr<StateRepresentation::Parameter<Eigen::MatrixXd>> stiffness_;///< stiffness matrix of the controller associated to position
+  std::shared_ptr<StateRepresentation::Parameter<Eigen::MatrixXd>> damping_;  ///< damping matrix of the controller associated to velocity
+  std::shared_ptr<StateRepresentation::Parameter<Eigen::MatrixXd>> inertia_;  ///< inertia matrix of the controller associated to acceleration
+
+public:
+  /**
+   * @brief Constructor initializing all the matrices
+   * @param stiffness the stiffness matrix
+   * @param damping the damping matrix
+   * @param inertia the inertia matrix
+   */
+  explicit Impedance(const Eigen::MatrixXd& stiffness, const Eigen::MatrixXd& damping, const Eigen::MatrixXd& inertia);
+
+  /**
+   * @brief Compute the force (task space) or torque (joint space) command based on the input state 
+   * of the system as the error between the desired state and the real state.
+   * @param desired_state the desired state to reach
+   * @param feedback_state the real state of the system as read from feedback loop
+   * @return the output command at the input state
+   */
+  virtual S compute_command(const S& desired_state, const S& feedback_state) const;
+
+  /**
+   * @brief Getter of the stiffness matrix
+   * @return the stiffness matrix
+   */
+  const Eigen::MatrixXd& get_stiffness() const;
+
+  /**
+   * @brief Getter of the dmaping matrix
+   * @return the damping matrix
+   */
+  const Eigen::MatrixXd& get_damping() const;
+
+  /**
+   * @brief Getter of the inertia matrix
+   * @return the inertia matrix
+   */
+  const Eigen::MatrixXd& get_inertia() const;
+
+  /**
+   * @brief Setter of the stiffness matrix
+   * @param the new stiffness matrix
+   */
+  void set_stiffness(const Eigen::MatrixXd& stiffness);
+
+  /**
+   * @brief Setter of the damping matrix
+   * @param the new damping matrix
+   */
+  void set_damping(const Eigen::MatrixXd& damping);
+
+  /**
+   * @brief Setter of the inertia matrix
+   * @param the new inertia matrix
+   */
+  void set_inertia(const Eigen::MatrixXd& inertia);
+
+  /**
+   * @brief Return a list of all the parameters of the controller
+   * @return the list of parameters
+   */
+  virtual std::list<std::shared_ptr<StateRepresentation::ParameterInterface>> get_parameters() const;
+};
+
+template <class S>
+Impedance<S>::Impedance(const Eigen::MatrixXd& stiffness, const Eigen::MatrixXd& damping, const Eigen::MatrixXd& inertia) : stiffness_(std::make_shared<StateRepresentation::Parameter<Eigen::MatrixXd>>("stiffness", stiffness)),
```

Comment:
Can the constructor validate the template type and throw an unsupported exception if not joint state / cartesian state? Or will this instead be handled by base "not implemented" exception? (if so that's a pretty interesting consequence for using that over pure virtual functions - you can always compile but might fail at runtime)

Reply:
There is one possibility to do that is to define all the accepted implementations in the `.cpp` file. This way, the compiler will fail to compile with anything else. It is a bit tedious if you have many accepted classes (hello `Parameters`) but quite robust.